src/persist-client/src/read.rs

```diff
@@ -1188,14 +1206,16 @@ where
     D: Semigroup + Codec64 + Send + Sync,
 {
     fn drop(&mut self) {
-        if let Some(heartbeat_tasks) = self.heartbeat_tasks.take() {
-            for heartbeat_task in heartbeat_tasks {
-                heartbeat_task.abort();
-            }
-        }
-        if self.explicitly_expired {
+        if self.unexpired_state.is_none() {
             return;
         }
+
+        // We drop the unexpired state before expiring the reader to ensure the
+        // heartbeat tasks can never observe the expired state. This doesn't
+        // matter for correctness, but avoids confusing log output if the
+        // heartbeat task were to discover that its lease has been expired.
```

Comment:
I _think_ its ok? The `abort` will take effect at the next `await` point. So I don't _think_ there's an interleaving that results in the task observing the results of the lease expiration.

Reply:
Sounds good! I wrote up a comment explaining why we warn loudly if the heartbeat task exits but _don't_ panic, because we're not 100% sure that the synchronization works out: https://github.com/MaterializeInc/materialize/pull/22637/files#diff-a785136af646da6a0d2adf114779fe78f36c1af754f76b393f97b51fe5805cedR1077-R1083

Thanks for the review!